docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,424 @@
+    case len(denomSplit) == 2:
+      err = tmtypes.ValidateHash([]byte(denomSplit[1]))
+    default:
+      err = Wrap(ErrInvalidDenomForTransfer, denom)
+  }
+
+  if err != nil {
+    return nil, err
+  }
+
+  return denomSplit[1], nil
+}
+```
+
+When a fungible token is sent to from a source chain, the trace information needs to be stored with the
+new port and channel identifiers:
+
+```golang
+// SendTransfer
+// ...
+
+// NOTE: SendTransfer simply sends the denomination as it exists on its own
+// chain inside the packet data. The receiving chain will perform denom
+// prefixing as necessary.
+
+if types.SenderChainIsSource(sourcePort, sourceChannel, token.Denom) {
+  // create the escrow address for the tokens
+  escrowAddress := types.GetEscrowAddress(sourcePort, sourceChannel)
+
+  // escrow source tokens. It fails if balance insufficient.
+  if err := k.bankKeeper.SendCoins(
+    ctx, sender, escrowAddress, sdk.NewCoins(token),
+  ); err != nil {
+    return err
+  }
+} else {
+  // set the value to the lookup table if not stored already
+  traceHash := denomTrace.Hash()
+  if !k.HasDenomTrace(ctx, traceHash) {
+    k.SetDenomTrace(ctx, traceHash, denomTrace)
+  }
+
+  // transfer the coins to the module account and burn them
+  if err := k.bankKeeper.SendCoinsFromAccountToModule(
+    ctx, sender, types.ModuleName, sdk.NewCoins(token),
+  ); err != nil {
+    return err
+  }
+
+  if err := k.bankKeeper.BurnCoins(
+    ctx, types.ModuleName, sdk.NewCoins(token),
+  ); err != nil {
+    // NOTE: should not happen as the module account was
+    // retrieved on the step above and it has enough balance
+    // to burn.
+    return err
+  }
+}
+// ...
+```
+
+The denomination trace info also needs to be updated when token is received in both cases:
+
+- Sender is **sink** chain: Store the received denomination, i.e in the [example](#example) above,
+  during step 4, when chain `C` receives the `A/C/B/A/denom`. As there the trimmed trace info is
+  already known by the chain we don't need to store it (i.e `C/B/A/denom`).
+- Sender is **source** chain: Store the received info. For example, during step 1, when chain `B` receives `A/denom`.
+
+```golang
+// OnRecvPacket
+// ...
+
+// This is the prefix that would have been prefixed to the denomination
+// on sender chain IF and only if the token originally came from the
+// receiving chain.
+//
+// NOTE: We use SourcePort and SourceChannel here, because the counterparty
+// chain would have prefixed with DestPort and DestChannel when originally
+// receiving this coin as seen in the "sender chain is the source" condition.
+voucherPrefix := types.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+
+if types.ReceiverChainIsSource(voucherPrefix, data.Denom) {
+  // sender chain is not the source, unescrow tokens
+
+  // set the value to the lookup table if not stored already
+  traceHash := denomTrace.Hash()
+  if !k.HasDenomTrace(ctx, traceHash) {
```

Comment:
since the receiver is the source, the receiver minted and escrowed the tokens so we can assume the prefix is set

Reply:
I think this check can be removed. We just need to check once if the lookup exists, that'd be upon a recv when the receiver is not the source.